core/src/main/java/org/cru/contentscoring/core/servlets/CopyScoresToTagsServlet.java

```diff
@@ -0,0 +1,259 @@
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.felix.scr.annotations.sling.SlingServlet;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.SlingHttpServletResponse;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.servlets.SlingAllMethodsServlet;
+import org.apache.sling.settings.SlingSettingsService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import java.io.IOException;
+import java.security.Principal;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+
+import static org.cru.contentscoring.core.service.impl.SyncScoreServiceImpl.SCALE_OF_BELIEF_TAG_PREFIX;
+
+@SlingServlet(
+    paths = "/bin/cru/content-scoring/move-scores-to-tags",
+    metatype = true,
+    methods = {"PUT"}
+)
+public class CopyScoresToTagsServlet extends SlingAllMethodsServlet {
+    private static final Logger LOG = LoggerFactory.getLogger(CopyScoresToTagsServlet.class);
+
+    private static final String PRIMARY_XF_NAME = "primaryExperienceFragment";
+    private static final String XF_TYPE = "cq/experience-fragments/components/experiencefragment";
+    private static final String XF_VARIANT_TYPE = "cq:xfVariantType";
+
+    @Reference
+    private SlingSettingsService slingSettingsService;
+
+    @Reference
+    private Replicator replicator;
+
+    @Override
+    protected void doPut(final SlingHttpServletRequest request, final SlingHttpServletResponse response)
+        throws IOException {
+
+        String path = request.getParameter("path");
+        if (Strings.isNullOrEmpty(path)) {
+            response.sendError(400, "Path is required");
+            return;
+        }
+
+        Principal principal = request.getUserPrincipal();
+        if (principal == null || !principal.getName().equals("admin")) {
+            LOG.error(
+                "Unauthorized attempt to move scores to tags by {}",
+                principal == null ? "Anonymous" : principal.getName());
+            response.sendError(401, "You are not authorized to perform this command.");
+            return;
+        }
+
+        ResourceResolver resourceResolver = request.getResourceResolver();
+        Resource root = resourceResolver.getResource(path);
+        if (root == null) {
+            response.sendError(400, "Invalid path");
+            return;
+        }
+
+        TagManager tagManager = resourceResolver.adaptTo(TagManager.class);
+
+        List<Hit> results = findPagesWithScoreProperty(root);
+
+        try {
+            moveScoresToTags(results, tagManager, resourceResolver);
+        } catch (RepositoryException | ReplicationException e) {
+            LOG.error(e.getMessage());
```

Comment:
normally it is helpful to capture the stacktrace, not just the message. Did you do this intentionally?

Reply:
I didn't bother since this code will all go away once I run the process in production, but I can capture it in case things go wrong.